src/pybind/mgr/progress/module.py

```diff
@@ -167,20 +177,27 @@ def pg_update(self, pg_dump, log):
             if "active" in states and "clean" in states and not unmoved:
                 complete.add(pg)
             else:
-                # FIXME: handle apparent negative progress
-                #   (e.g. if num_bytes_recovered goes backwards)
-                # FIXME: handle apparent >1.0 progress
-                #   (num_bytes_recovered incremented by more than num_bytes)
                 if info['stat_sum']['num_bytes'] == 0:
                     # Empty PGs are considered 0% done until they are
                     # in the correct state.
                     pass
                 else:
                     recovered = info['stat_sum']['num_bytes_recovered']
                     total_bytes = info['stat_sum']['num_bytes']
-                    ratio = float(recovered -
-                                  self._original_bytes_recovered[pg]) / \
-                        total_bytes
+                    if total_bytes > 0:
+                        ratio = float(recovered -
+                                      self._original_bytes_recovered[pg]) / \
+                            total_bytes
+
+                        # Since the recovered bytes (over time) could perhaps
+                        # exceed the contents of the PG (moment in time), we
+                        # must clamp this
+                        ratio = min(ratio, 1.0)
+
+                    else:
+                        # Dataless PGs (e.g. containing only OMAPs) count
+                        # as half done.
+                        ratio = 0.5
```

Comment:
If bytes aren't (currently) available as a metric for omap data, would it be possible to substitute a separate metric like the number of objects recovered? Are there any applications that store a significant amount of data in omap?